File: edition4/js/ch02/2-3.js
function Node(data, next) {
  this.data = data;
  this.next = next;
};

function removeNode(n) {
  if (!n || !n.next) {
    return false;
  } else {
    n.data = n.next.data;
    n.next = n.next.next;
    return true;
  }
}

// tests
var assert = require('assert');

describe('removeNode', function() {

  it('should fail when n is falsy', function() {
    assert.equal(removeNode(null), false);
  });

  it('should fail when n.next is falsy', function() {
    var n = new Node(1, null);
    assert.equal(removeNode(n), false);
  });

  it('should succeed when n is the first node in a list', function() {
    var n2 = new Node('3', null);
    var n1 = new Node('2', n2);
    var head = new Node('1', n1);
    assert.equal(removeNode(head), true);
    assert.equal(head.data, n1.data);
    assert.equal(head.next, n1.next);
  });

  it('should succeed when n is in the middle of a list', function() {
    var n3 = new Node('4', null);
    var n2 = new Node('3', n3);
    var n1 = new Node('2', n2);
    var head = new Node('1', n1);
    assert.equal(removeNode(n1), true);
    assert.equal(n1.data, n2.data);
    assert.equal(n1.next, n2.next);
  });

  it('should succeed when n is the second to last in a list', function() {
    var n3 = new Node('4', null);
    var n2 = new Node('3', n3);
    var n1 = new Node('2', n2);
    var head = new Node('1', n1);
    assert.equal(removeNode(n2), true);
    assert.equal(n2.data, n3.data);
    assert.equal(n2.next, n3.next);
  });

});

File: edition4/c/ch02/2-3.cpp
#include <assert.h>

#include <algorithm>
#include <cmath>
#include <iostream>
#include <string>

using namespace std;

class Node {
  public:
    int value;
    Node *next;
    Node(int value, Node *next);
};

Node::Node(int value, Node *next) {
  this->value = value;
  this->next = next;
}

class LinkedList {
  public:
    Node *head;
    LinkedList();
    void add(int value);
    void deleteFromNode(Node *n);
    int length();
    void print();
};

LinkedList::LinkedList() {
  this->head = NULL;
}

void LinkedList::add(int value) {
  if (!this->head) {
    this->head = new Node(value, NULL);
  } else {
    Node *temp = this->head;

    while (temp->next) {
      temp = temp->next;
    }

    temp->next = new Node(value, NULL);
  }
}

void LinkedList::deleteFromNode(Node *n) {
  if (!n || !n->next) {
    return;
  }

  Node *next = n->next;

  n->value = n->next->value;
  n->next = n->next->next;

  delete next;
}

int LinkedList::length() {
  int length = 0;
  Node *curr = this->head;

  while (curr) {
    length++;
    curr = curr->next;
  }

  return length;
}

void LinkedList::print() {
  Node *curr = this->head;

  while (curr) {
    cout << curr->value << " -> ";
    curr = curr->next;
  }

  cout << "NULL" << endl;
}

int main() {
  LinkedList list;
  list.deleteFromNode(list.head);

  list.add(1);
  list.add(2);
  list.add(3);
  list.add(4);
  list.deleteFromNode(list.head->next);
  assert(list.length() == 3);
  assert(list.head->value == 1);
  assert(list.head->next->value == 3);
  assert(list.head->next->next->value == 4);

  return 0;
}
